File: apps/sdkwork-webserver-flutter-mobile/packages/sdkwork_webserver_flutter_mobile_commons/lib/src/widgets/screen_states.dart
import 'package:flutter/material.dart';

import '../theme/design_tokens.dart';

/// Domain-neutral screen/list state primitives.
///
/// `FLUTTER_APP_MOBILE_ARCHITECTURE_SPEC.md` section 4 — commons owns the shared
/// primitives, capability packages own the pages and map their own payloads onto
/// these payload-free widgets.
enum WebserverFlutterScreenStatus { loading, ready, empty, error }

/// Collapse an item count, an in-flight flag, and an error message into the one
/// status every list screen renders.
///
/// Pure so the contract test can exercise the state machine without a widget
/// binding; it mirrors `resolveWebserverHarmonyScreenStatus` and the H5/mini
/// program resolvers.
WebserverFlutterScreenStatus resolveWebserverFlutterScreenStatus(
  int itemCount,
  bool loading,
  String errorMessage,
) {
  if (loading) {
    return WebserverFlutterScreenStatus.loading;
  }
  if (errorMessage.isNotEmpty) {
    return WebserverFlutterScreenStatus.error;
  }
  return itemCount == 0
      ? WebserverFlutterScreenStatus.empty
      : WebserverFlutterScreenStatus.ready;
}

/// Message key a given status renders. Keeping the mapping here means a new
/// status cannot be added without deciding what it says.
String resolveWebserverFlutterScreenMessageKey(
  WebserverFlutterScreenStatus status,
) {
  switch (status) {
    case WebserverFlutterScreenStatus.loading:
      return 'applications.list.loading';
    case WebserverFlutterScreenStatus.error:
      return 'applications.list.error';
    case WebserverFlutterScreenStatus.empty:
      return 'applications.list.empty';
    case WebserverFlutterScreenStatus.ready:
      return '';
  }
}

/// Payload-free status view used by capability screens.
class WebserverFlutterStatusView extends StatelessWidget {
  const WebserverFlutterStatusView({super.key, required this.message});

  final String message;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Padding(
        padding: const EdgeInsets.all(WebserverFlutterTokens.spacingLg),
        child: Text(
          message,
          textAlign: TextAlign.center,
          style: const TextStyle(
            color: Color(WebserverFlutterTokens.colorTextMuted),
          ),
        ),
      ),
    );
  }
}

File: apps/sdkwork-webserver-mini-program/packages/sdkwork-webserver-mp-commons/src/components/screenStates.ts
/**
 * Domain-neutral screen/list state primitives.
 *
 * Capability packages map these to their own data; the primitives stay
 * payload-free so they remain reusable across capabilities
 * (`MINI_PROGRAM_APP_ARCHITECTURE_SPEC.md` §3 — commons owns the shared
 * primitives, capability packages own the pages).
 */
export type WebserverMpScreenStatus = "loading" | "ready" | "empty" | "error";

export interface WebserverMpScreenState {
  readonly status: WebserverMpScreenStatus;
  readonly errorMessage?: string;
}

export const initialWebserverMpScreenState: WebserverMpScreenState = { status: "loading" };

export function resolveWebserverMpScreenStatus(
  itemCount: number,
  loading: boolean,
  errorMessage?: string,
): WebserverMpScreenStatus {
  if (loading) {
    return "loading";
  }
  if (typeof errorMessage === "string" && errorMessage.length > 0) {
    return "error";
  }
  return itemCount === 0 ? "empty" : "ready";
}

export function resolveWebserverMpScreenState(
  itemCount: number,
  loading: boolean,
  errorMessage?: string,
): WebserverMpScreenState {
  return {
    status: resolveWebserverMpScreenStatus(itemCount, loading, errorMessage),
    ...(errorMessage ? { errorMessage } : {}),
  };
}
